services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/serialization/ResumableFileUploadSerializer.java

```diff
@@ -0,0 +1,180 @@
+        jsonGenerator.writeEndObject();
+
+        return jsonGenerator.getBytes();
+    }
+
+    /**
+     * At this point we do not need to persist the TransferRequestOverrideConfiguration, because it only contains listeners and
+     * they are not used in the resume operation.
+     */
+    private static void marshallUploadFileRequest(UploadFileRequest fileRequest, JsonWriter jsonGenerator) {
+        jsonGenerator.writeFieldName(UPLOAD_FILE_REQUEST);
+        jsonGenerator.writeStartObject();
+        jsonGenerator.writeFieldName(SOURCE);
+        jsonGenerator.writeValue(fileRequest.source().toString());
+        marshallPutObjectRequest(fileRequest.putObjectRequest(), jsonGenerator);
+        jsonGenerator.writeEndObject();
+    }
+
+    private static void marshallPutObjectRequest(PutObjectRequest putObjectRequest, JsonWriter jsonGenerator) {
+        jsonGenerator.writeFieldName(PUT_OBJECT_REQUEST);
+        jsonGenerator.writeStartObject();
+        validateNoRequestOverrideConfiguration(putObjectRequest);
+        putObjectRequest.sdkFields().forEach(field -> marshallPojoField(field, putObjectRequest, jsonGenerator));
+        jsonGenerator.writeEndObject();
+    }
+
+    private static void validateNoRequestOverrideConfiguration(PutObjectRequest putObjectRequest) {
+        if (putObjectRequest.overrideConfiguration().isPresent()) {
+            log.debug(() -> "ResumableFileUpload PutObjectRequest contains an override configuration that will not be "
+                            + "serialized");
+        }
+    }
+
+    private static void marshallPojoField(SdkField<?> field, PutObjectRequest request, JsonWriter jsonGenerator) {
+        Object val = field.getValueOrDefault(request);
+        TransferManagerJsonMarshaller<Object> marshaller = getMarshaller(field.marshallingType(), val);
+        marshaller.marshall(val, jsonGenerator, field.locationName());
+    }
+
+    public static ResumableFileUpload fromJson(String bytes) {
+        JsonNodeParser jsonNodeParser = JsonNodeParser.builder().build();
+        Map<String, JsonNode> uploadNodes = jsonNodeParser.parse(bytes).asObject();
+        return fromNodes(uploadNodes);
+    }
+
+    public static ResumableFileUpload fromJson(byte[] bytes) {
+        JsonNodeParser jsonNodeParser = JsonNodeParser.builder().build();
+        Map<String, JsonNode> uploadNodes = jsonNodeParser.parse(bytes).asObject();
+        return fromNodes(uploadNodes);
+    }
+
+    public static ResumableFileUpload fromJson(InputStream bytes) {
+        JsonNodeParser jsonNodeParser = JsonNodeParser.builder().build();
+        Map<String, JsonNode> uploadNodes = jsonNodeParser.parse(bytes).asObject();
+        return fromNodes(uploadNodes);
+    }
+
+    private static ResumableFileUpload fromNodes(Map<String, JsonNode> uploadNodes) {
+        TransferManagerJsonUnmarshaller<Object> longUnmarshaller = getUnmarshaller(MarshallingType.LONG);
```

Comment:
Query : Curious why is this not ` TransferManagerJsonUnmarshaller<Long> ` so that we can avoid casting down the line

Reply:
Yeah, let me fix it